src/Http.php

```
@@ -0,0 +1,239 @@
+        }
+
+        if (isset($options['json'])) {
+            $body = json_encode($options['json']);
+            $headers['Content-Type'] = 'application/json';
+
+            return [$body, $headers];
+        }
+
+        return [$options['body'] ?? null, $headers];
+    }
+
+    /**
+     * Handle an asynchronous HTTP request.
+     *
+     * @param \GuzzleHttp\Client $client
+     * @param string             $method
+     * @param string             $url
+     * @param array              $requestOptions
+     *
+     * @return \GuzzleHttp\Promise\PromiseInterface
+     */
+    protected function handleAsyncRequest(
+        Client $client,
+        string $method,
+        string $url,
+        array $requestOptions
+    ): PromiseInterface {
+        return $client->requestAsync($method, $url, $requestOptions)->then(
+            fn (ResponseInterface $response) => new Response($response),
+            fn (RequestException $e) => $this->handleRequestException($e)
+        );
+    }
+
+    /**
+     * Handle a synchronous HTTP request.
+     *
+     * @param \GuzzleHttp\Client $client
+     * @param string             $method
+     * @param string             $url
+     * @param array              $requestOptions
+     *
+     * @return \Fetch\Response
+     */
+    protected function handleSyncRequest(
+        Client $client,
+        string $method,
+        string $url,
+        array $requestOptions
+    ): Response {
+        try {
+            $response = $client->request($method, $url, $requestOptions);
+
+            return new Response($response);
+        } catch (RequestException $e) {
+            return $this->handleRequestException($e);
+        }
+    }
+
+    /**
+     * Handle a request exception and return a mock response.
+     *
+     * @param \GuzzleHttp\Exception\RequestException $e
+     *
+     * @return \Fetch\Response
+     */
+    protected function handleRequestException(RequestException $e): Response
```

returning a `Response` when an exception occurs is an anti-pattern. It's better to throw a custom exception class and let the users of the package handle it.